abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+`proxy_on_new_connection` and `proxy_on_http_request_headers` callbacks.
+
+Return value (`new_context_id`) can be used to assign new context identifier
+that will be used by the host environment in subsequent calls. Returning `0`
+retains the original context identifier.
+
+#### `proxy_on_context_finalize`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - `i32 (bool) is_done`
+
+Called when the host environment is done processing context (`context_id`).
+Any finalization (e.g. logging) should be done within this callback.
+
+Return value (`is_done`) indicates that the Wasm module also finished processing
+context (`context_id`). If it didn't, then it should return `false` and call
+`proxy_context_finalize` when it's done.
+
+### Functions exposed by the host environment
+
+#### `proxy_set_effective_context`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Change the effective context to `context_id`. This function is usually used to
+change effective context during `proxy_on_queue_ready`, `proxy_on_timer_ready`,
+`proxy_on_http_call_response` and `proxy_on_grpc_call_response` callbacks.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `context_id`.
+
+#### `proxy_context_finalize`
+
+* params:
+  - none
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Indicates to the host environment that the Wasm module finished processing
+current context. It should be used after previously returning `false` in
+`proxy_on_context_finalize`.
+
+Returns `Ok` on success.
+
+
+## Configuration
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_vm_start`
+
+* params:
+  - `i32 (uint32_t) vm_id`
+  - `i32 (size_t) vm_configuration_size`
+* returns:
+  - `i32 (bool) is_success`
+
+Called when the host environment starts WebAssembly Virtual Machine (WasmVm).
+
+Its configuration (of size `vm_configuration_size`) can be retrieved using
+`proxy_get_buffer`.
+
+Return value (`is_success`) indicates that configuration was processed correctly
+and that WasmVm started successfully.
+
+#### `proxy_on_plugin_start`
+
+* params:
+  - `i32 (uint32_t) plugin_id`
+  - `i32 (size_t) plugin_configuration_size`
+* returns:
+  - `i32 (bool) is_success`
+
+Called when the host environment starts plugin.
+
+Its configuration (of size `plugin_configuration_size`) can be retrieved using
+`proxy_get_buffer`.
+
+Return value (`is_success`) indicates that configuration was processed correctly
+and that plugin started successfully.
+
+
+## TCP/UDP/QUIC stream (L4) extensions
+
+Note: `downstream` means the connection between client and proxy, and `upstream`
+       means the connection between proxy and backend (aka “next hop”).
+
+Note: The same unique context identifier (`stream_id`) is shared by a pair of
+      `downstream` and `upstream` connections.
+
+### Callbacks optionally exposed by the Wasm module
+
+#### `proxy_on_new_connection`
```

Comment:
Can you specify the state machine of the callbacks? E.g. 
```regex
new_connection (downstream_data | upstream_data)* (downstream_close upstream_data* upstream_close | upstream_close downstream_data* downstream_close)
```

Also, specify the interleavings with configuration callbacks.